### ISSUE TITLE
Allow veto of certain combinations

### ISSUE THREAD

**Issue by AArnott** [AArnott/Xunit.Combinatorial#3]
`PairwiseDataAttribute` and `CombinatorialDataAttribute` should accept a parameter that allows the test to veto certain known invalid parameter combinations. Thereby suppressing generation of test cases that don't make sense.

Ideally in the pairwise case, such a suppression will feed back into the pairwise algorithm so it can try selecting other combinations that still satisfy pairwise testing requirements, wherever possible.

**Comment by ilyabreev** [comment]
+1 for this feature. It seems not really hard to implement it at least for `CombinatorialDataAttribute`.

**Comment by alrz** [comment]
+1

Currently we have to do `switch (param1, param2) { case (invalid, combinations): .. return; }` which doesn't prevent redundant test runs.

**Comment by AArnott** [comment]
Can someone make a proposal that includes the new API surface area, example uses and describe the behavior?

**Comment by alrz** [comment]
This is the simplest I could come up with
```cs
[Theory, CombinatorialData]
[CombinatiorialExclude(false, false, false)]
// maybe a few more
public void CheckValidAge(bool p1, bool p2, bool p3)

```
to exclude `(false, false, false)` test case.

Something similar should probably work for pairwise, but I don't have a good use case for it right now.

**Comment by AArnott** [comment]
That's a good start. What about when the args can't be expressed as constants? Something like MemberData provides. Or maybe that already doesn't work for Combinatorial, I don't recall.

**Comment by alrz** [comment]
>What about when the args can't be expressed as constants? Something like MemberData provides. Or maybe that already doesn't work for Combinatorial,

Even if it does work, that may need a special equality check as well so it's probably out of scope.

I think an attribute like that will cover 90% of use cases, I wouldn't do anything further until there's a definite use case for it.

**Comment by AArnott** [comment]
#1 actually tracks a `MemberData`-like attribute and it looks like we may have a volunteer for contributing it soon.

**Comment by alrz** [comment]
https://github.com/AArnott/Xunit.Combinatorial/commit/26696c117d5f65dbcdf91209ae0b0816593cec4d doesn't seem to be related?

**Comment by AArnott** [comment]
Agreed. That's a mistake from merging commits from another repo. Thanks for catching, @alrz.

**Comment by relfman-cmg** [comment]
+1

**Comment by snake-scaly** [comment]
I just stumbled upon the need for something like `[CombinatorialSkip(null, null, Any())]`, that is, to ignore all combinations where the first two parameters are `null`. In my particular case, #93 would also work.

**Comment by AArnott** [comment]
These data-attributes don't mean this library *runs* the test -- xunit default test runner still does. But I wonder if xunit defines an exception type that can be thrown from a test to say "this test case is invalid. skip it". In which case, we wouldn't be limited to attributes to define which combinations are invalid. 

It might be very useful to be able to do something like this in the test method itself:

```cs
Skip.When(a1 % 3 > a3, "Invalid test case");
```

In fact that particular syntax probably _would_ work right now since Xunit v3 has a `SkipException`. So you all could use that today.

Having attributes from this library above and beyond that would provide a limited ability to skip very specific cases such that they aren't even enumerated and won't show as 'skipped' in the test results.

**Comment by snake-scaly** [comment]
For now I ended up using your SkippableTheory and the `Skip.If` predicate. Unfortunately it doesn't play well with Pairwise, and I suspect your exception idea won't either.
